# References

- Last modified: ons apr 03, 2024  10:54
- Sign: JN

## Citations and links to software

- [ASTER](https://github.com/chaoszhang/ASTER) Zhang, C., and S. Mirarab. 2022.
  Weighting by Gene Tree Uncertainty Improves Accuracy of Quartet-based Species
  Trees. Mol. Biol. Evol. 39:msac215. <https://doi.org/10.1093/molbev/msac215>
- [ASTRAL](https://github.com/smirarab/ASTRAL) Zhang, C., M. Rabiee, E.
  Sayyari, and S. Mirarab. 2018. ASTRAL-III: Polynomial Time Species Tree
  Reconstruction from Partially Resolved Gene Trees. BMC Bioinformatics
  19(S6):153. <https://doi.org/10.1186/s12859-018-2129-y>
- [ATPW](https://github.com/nylander/Align-and-trees-parallel-workflow)
  Nylander, J. A. A. 2022. ATPW - Align and trees parallel workflow. Program
  distributed by the author.
  <https://github.com/nylander/Align-and-trees-parallel-workflow>
- [BMGE](http://ftp.pasteur.fr/pub/gensoft/projects/BMGE/) Criscuolo, A. and S.
  Gribaldo. 2010. BMGE (Block Mapping and Gathering with Entropy): a new
  software for selection of phylogenetic informative regions from multiple
  sequence alignments. BMC Evol. Biol. 10:210.
  <https://doi.org/10.1186/1471-2148-10-210>
- [FastaGap](https://github.com/nylander/fastagap) Nylander, J. A. A. 2019.
  FastaGap. Software distributed by the author.
  <https://github.com/nylander/fastagap>
- [GNU parallel](https://www.gnu.org/software/parallel) Tange, O. 2021. GNU
  Parallel 20210822 ('Kabul'). Zenodo. <https://doi.org/10.5281/zenodo.5233953>
- [MAFFT](https://gitlab.com/sysimm/mafft) Katoh, K. and D. M.
  Standley. 2013. MAFFT Multiple Sequence Alignment Software Version 7:
  Improvements in Performance and Usability. Mol. Biol. Evol. 30:772-780.
  <https://doi.org/10.1093/molbev/mst010>
- [ModelTest-NG](https://github.com/ddarriba/modeltest) Darriba, D., D. Posada,
  A. M. Kozlov, A. Stamatakis, B. Morel, and T. Flouri. 2020. ModelTest-NG: a
  new and scalable tool for DNA and protein model selection. Mol. Biol. Evol.
  37:291-294. <https://doi.org/10.1093/molbev/msz189>
- [ParGenes](https://github.com/BenoitMorel/ParGenes) Benoit Morel, B., A. M.
  Kozlov, and A. Stamatakis. 2018. ParGenes: a tool for massively parallel
  model selection and phylogenetic tree inference on thousands of genes.
  Bioinformatics 35:1771-1773. <https://doi.org/10.1093/bioinformatics/bty839>
- [RaxML-NG](https://github.com/amkozlov/raxml-ng) Kozlov, A. M., D. Darriba,
  T. Flouri, B. Morel, and A. Stamatakis. 2019. RAxML-NG: A fast, scalable, and
  user-friendly tool for maximum likelihood phylogenetic inference.
  Bioinformatics 35:4453-4455. <https://doi.org/10.1093/bioinformatics/btz305>
- [TreeShrink](https://github.com/uym2/TreeShrink) Mai, U., and S. Mirarab.
  2018. TreeShrink: Fast and Accurate Detection of Outlier Long Branches in
  Collections of Phylogenetic Trees. BMC Genomics 19(S5):272.
  <https://doi.org/10.1186/s12864-018-4620-2>
- [TrimAL](https://github.com/inab/trimal) Capella-Gutierrez, S., J. M.
  Silla-Martinez, and T. Gabaldon. 2009. trimAl: a tool for automated alignment
  trimming in large-scale phylogenetic analyses. Bioinformatics 25:1972-1973.
  <https://doi.org10.1093/bioinformatics/btp348>

